handler/src/main/java/io/netty/handler/ssl/Java9SslUtils.java

```diff
@@ -0,0 +1,90 @@
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.ssl;
+
+import io.netty.util.internal.PlatformDependent;
+
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLParameters;
+import java.lang.reflect.Method;
+import java.util.List;
+
+final class Java9SslUtils {
+    private static final Method setApplicationProtocols;
+    private static final Method getApplicationProtocol;
+
+    static {
+        Method setter = null;
+        Method getter = null;
+
+        if (PlatformDependent.javaVersion() >= 9) {
+            try {
+                getter = SSLEngine.class.getMethod("getApplicationProtocol");
+                setter = SSLParameters.class.getMethod("setApplicationProtocols", String[].class);
+            } catch (Throwable t) {
+                throw new RuntimeException("Unable to initialize Java9SslUtils", t);
+            }
+        }
+
+        getApplicationProtocol = getter;
+        setApplicationProtocols = setter;
+    }
+
+    private Java9SslUtils() {
+    }
+
+    static boolean supportsAlpn(SSLEngine engine) {
+        if (PlatformDependent.javaVersion() < 9) {
+            return false;
+        }
+
+        try {
+            getApplicationProtocol.invoke(engine);
+            return true;
+        } catch (Exception ignore) {
+            return false;
+        }
+    }
+
+    static SSLEngine configureAlpn(SSLEngine engine, JdkApplicationProtocolNegotiator applicationNegotiator) {
+        List<String> supportedProtocols = applicationNegotiator.protocols();
+        SSLParameters params = engine.getSSLParameters();
+        try {
+            setApplicationProtocols(params, supportedProtocols);
+        } catch (Exception ex) {
+            throw new RuntimeException(ex);
+        }
+        engine.setSSLParameters(params);
+        return engine;
+    }
+
+    static String getApplicationProtocol(SSLEngine sslEngine) {
+        try {
+            return (String) getApplicationProtocol.invoke(sslEngine);
+        } catch (Exception ex) {
+            throw new RuntimeException(ex);
+        }
+    }
+
+    private static void setApplicationProtocols(SSLParameters parameters, List<String> supportedProtocols) {
+        String[] protocolArray = supportedProtocols.toArray(new String[]{});
+        try {
+            setApplicationProtocols.invoke(parameters, new Object[]{ protocolArray });
+        } catch (Exception ex) {
```

Comment:
http://docs.oracle.com/javase/7/docs/api/java/lang/Class.html#getMethod(java.lang.String,%20java.lang.Class...) on the other hand may throw a SecurityException, nice catch @normanmaurer!